src/main/java/org/folio/edge/sip2/repositories/PatronRepository.java

```diff
@@ -89,22 +95,84 @@
     final Future<JsonObject> result = usersRepository.getUserByBarcode(barcode, sessionData);
     return result.compose(user -> {
       if (user == null || !user.getBoolean("active", FALSE).booleanValue()) {
-        return invalidPatron(patronInformation);
+        return invalidPatron(patronInformation, null);
       } else {
         final String userId = user.getString("id");
         if (userId == null) {
           // Something is really messed up if the id is missing
           log.error("User with barcode {} is missing the \"id\" field", barcode);
-          return invalidPatron(patronInformation);
+          return invalidPatron(patronInformation, null);
         }
-        return validPatron(userId, user.getJsonObject("personal", new JsonObject()),
-            patronInformation, sessionData);
+
+        final Future<IResource> loginFuture;
+        final boolean patronPasswordVerificationRequired =
+            sessionData.isPatronPasswordVerificationRequired();
+        if (patronPasswordVerificationRequired) {
+          loginFuture = loginRepository.patronLogin(user.getString("username"),
+              patronInformation.getPatronPassword(),
+              sessionData);
+        } else {
+          loginFuture = Future.succeededFuture();
+        }
+
+        return loginFuture.compose(login -> {
+          // This means we tried to log in as the patron and failed in some way
+          if (login != null && !login.getErrorMessages().isEmpty()) {
+            return invalidPatron(patronInformation, FALSE);
+          } else {
+            return validPatron(userId, user.getJsonObject("personal", new JsonObject()),
+                patronInformation, sessionData, patronPasswordVerificationRequired
+                && patronInformation.getPatronPassword() != null ? TRUE : null);
+          }
+        });
       }
     });
   }
 
+  /**
+   * Perform End Patron Session.
+   * If PIN verification is required, then we will ensure that the patron password (PIN) is
+   * validated before allowing a successful return.
+   * 
+   * @param endPatronSession command data
+   * @param sessionData session data
+   * @return the end session response indicating whether or not the session was ended
+   */
+  public Future<EndSessionResponse> performEndPatronSessionCommand(
+      EndPatronSession endPatronSession,
```

Comment:
In the ideal world, yes, this is redundant and a possibly stupid thing to do here. But, since SIP2 is trying to be stateless, by passing data in each command that should not require another command before it, I was being pedantic here by requiring a password verification before end session, in case, somehow, the password sent was changed to something that doesn't match the patron identifier. We may decide this is unnecessary at some point and remove it.

For now, anywhere there is a patron password to execute a SIP2 command, I think we should verify it if password verification is enabled.